src/charm.py

```diff
@@ -0,0 +1,241 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""A Juju charm for Blackbox Exporter."""
+
+import logging
+import socket
+from urllib.parse import urlparse
+
+import yaml
+from blackbox import ConfigUpdateFailure, WorkloadManager
+from charms.grafana_k8s.v0.grafana_dashboard import GrafanaDashboardProvider
+from charms.loki_k8s.v0.loki_push_api import LokiPushApiConsumer
+from charms.observability_libs.v0.kubernetes_compute_resources_patch import (
+    K8sResourcePatchFailedEvent,
+    KubernetesComputeResourcesPatch,
+    ResourceRequirements,
+    adjust_resource_requirements,
+)
+from charms.prometheus_k8s.v0.prometheus_scrape import MetricsEndpointProvider
+from ops.charm import ActionEvent, CharmBase
+from ops.main import main
+from ops.model import (
+    ActiveStatus,
+    BlockedStatus,
+    MaintenanceStatus,
+    WaitingStatus,
+)
+from ops.pebble import PathError, ProtocolError
+
+logger = logging.getLogger(__name__)
+
+
+class BlackboxExporterCharm(CharmBase):
+    """A Juju charm for Blackbox Exporter."""
+
+    # Container name must match metadata.yaml
+    # Service name matches charm name for consistency
+    _container_name = _service_name = "blackbox"
+    _port = 9115
+
+    # path, inside the workload container, to the blackbox exporter configuration files
+    _config_path = "/etc/blackbox_exporter/config.yml"
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        self.container = self.unit.get_container(self._container_name)
+
+        # Core lifecycle events
+        self.blackbox_workload = WorkloadManager(
+            self,
+            container_name=self._container_name,
+            port=self._port,
+            web_external_url=self._external_url,
```

Comment:
Does the workload manager need the external url?

Reply:
I imagine ext. url would only be needed for ingress. If we do not plan having ingress  here then we don't need that arg probably.